### PR TITLE
Minimal update of lock file

### PR DESCRIPTION
**Pull request by NathanReb**
This PR adds a `--minimal-update` flag to `lock` which allows for updating the lock file, aiming at minimum changes in the dependencies.

When the flag is passed, the target lockfile (i.e. the one that it will write the generated lock file to) must already exists. It parsed and the list of packages extracted from it. It builds a map `name -> version` that is then passed to the solver. The map is used by the context to promote those version to the top of the preference sorted list of candidates for the package.
This result in different versions only getting picked if no solution exist with the version form the previous lock file. New packages are picked as they normally would and package that are not required anymore are also dropped without issues.

In short the preference order allows us to do everything we wanted from that feature making this a very easy to implement yet extremely useful feature!

### CODE REVIEW

**Comment by NathanReb** [comment]
Note that the documentation page I added is obviously incomplete but I think it's better to improve it iteratively than to wait to have a perfection version of it before merging it. I'm happy to add to it in a separate PR!

File: test/bin/minimal-update.t/run.t [hunk 1]
```diff
@@ -0,0 +1,68 @@
+We have a simple project with a single package which depends on
+packages a and b:
```
Comment (Leonidas-from-XIV) [inline test/bin/minimal-update.t/run.t:2]: Just a small thing: it would be nice if these were `a`, `b`, `c` in the readme, because sometimes it is hard to distinguish between the package `a` and the word "a".

File: lib/opam_solve.ml [hunk 2]
```diff
@@ -175,6 +178,22 @@ module Opam_monorepo_context (Base_context : BASE_CONTEXT) :
     in
     regular_versions @ avoid_versions
 
+  let prefer_version ~version versions =
```
Comment (Leonidas-from-XIV) [inline lib/opam_solve.ml:181]: `preferred_version` would be nicer here. Maybe even something like `let promote_version ~preferred_version candidate_versions =` as a analogue to `demote`.

File: lib/opam_solve.ml [hunk 3]
```diff
@@ -175,6 +178,22 @@ module Opam_monorepo_context (Base_context : BASE_CONTEXT) :
...
+    let rec aux acc l =
```
Comment (Leonidas-from-XIV) [inline lib/opam_solve.ml:182]: `aux` is a rather nondescript name and given the function is somewhat dense (it is not a loop) I would try to give it a better name. Also `l` could be replaced by `function`.

File: lib/opam_solve.ml [hunk 4]
```diff
@@ -175,6 +178,22 @@ module Opam_monorepo_context (Base_context : BASE_CONTEXT) :
...
+      match l with
+      | [] -> versions
+      | (v, opam) :: tl when OpamPackage.Version.equal v version ->
```
Comment (Leonidas-from-XIV) [inline lib/opam_solve.ml:185]:
```suggestion
      | ((version, _) as candidate) :: tl when OpamPackage.Version.equal version preferred_version ->
        candidate :: List.rev_append acc tl
```
feels more readable to me because it describes what the tuple is and comparing  `preferred_version` with `version` feels more understandable than comparing `version` with `v`.

File: lib/opam_solve.ml [hunk 5]
```diff
@@ -52,6 +53,7 @@ module Opam_monorepo_context (Base_context : BASE_CONTEXT) :
     require_dune : bool;
     opam_provided : OpamPackage.Name.Set.t;
     require_cross_compile : bool;
+    preferred_versions : OpamTypes.version OpamPackage.Name.Map.t option;
```
Comment (Leonidas-from-XIV) [inline lib/opam_solve.ml:56]: As mentioned in the other comments, maybe an empty map would be a cyclomatically less complex version.

I feel that there is no semantic difference in this case between no preferred versions in a map and no map - to me it feels like it expresses the same state ("no version preferred") but with the option there are now two ways to express this.

File: lib/opam_solve.ml [hunk 6]
```diff
@@ -201,10 +221,15 @@ module Opam_monorepo_context (Base_context : BASE_CONTEXT) :
           require_cross_compile
           && List.exists ~f:candidate_cross_compile candidates
         in
+        let preferred_version =
+          let open Option.O in
+          let* preferred_versions = preferred_versions in
```
Comment (Leonidas-from-XIV) [inline lib/opam_solve.ml:226]: I feel like this would be nicer simpler if it was just am empty map, saves one indirection and expresses the same semantically.

File: lib/opam_solve.ml [hunk 7]
```diff
@@ -201,10 +221,15 @@ module Opam_monorepo_context (Base_context : BASE_CONTEXT) :
...
+          OpamPackage.Name.Map.find_opt name preferred_versions
+        in
         filter_candidates ~allow_jbuilder ~must_cross_compile ~require_dune
           ~name candidates
         |> remove_opam_provided ~opam_provided
-        |> demote_candidates_to_avoid
+        |> sort_candidates_by_preference ~preferred_version
```
Comment (Leonidas-from-XIV) [inline lib/opam_solve.ml:232]: Is there a way to keep this as two steps, somehow? I feel like the double processing overhead is outweighted by the readability of having functions do one thing. Then `sort_candidates_by_preference` becomes basically just a quick pattern match.

File: cli/lock.ml [hunk 8]
```diff
@@ -532,6 +557,18 @@ let require_cross_compile =
     (fun x -> `Require_cross_compile x)
     Arg.(value & flag & info ~doc [ "require-cross-compile" ])
 
+let minimal_update =
+  let doc =
+    "Generate a lock file with minimum dependency changes compared to the \
+     previous lock file. It can remove or add packages based on the dependency \
+     specification in the local opam files but will not upgrade or downgrade \
+     previously locked packages unless strictly necessary.\n\
+     Require a lock file to exist in the target lock file path."
```
Comment (Leonidas-from-XIV) [inline cli/lock.ml:566]: I feel like specifying the mechanism how it does so e.g. "prefers to keep the versions as in the existing lock file unless strictly necessary" would make it feel a bit less magical - then it is clear how it tries to do minimal dependency changes without complicating the description too much.

What do you think?

**Review by Leonidas-from-XIV — COMMENTED** [review]
I am also pleasantly surprised how little changes (at least from the core logic part) it is. Hence I mostly looked at that to suggest some improvements. Glad to see that the implementation matches how my hunch it could be implemented, which is great since that means that the complexity increase of the code is very manageable :+1:  

The flags & parsing and tests look good. It's a bunch of code but essentially not-exciting code and the compiler does a good job of making sure this checks out.

File: lib/opam_solve.ml (hunk 3 above)
Comment (NathanReb) [reply lib/opam_solve.ml:182]: I personnaly quite dislike `function`, I prefer to declare all of my arguments so if you don't mind I'll stick to this.

The `aux` function here is just a recursive version of the main function with an extra `accumulator` arg. I find it to be a quite common pattern but I can rename it `recursive_prefer_version` if you'd like.

File: lib/opam_solve.ml (hunk 2 above)
Comment (NathanReb) [reply lib/opam_solve.ml:181]: `preferred_version` would be confusing here as it just re-orders the list and that name would somehow indicate it will return a version.

I initially thought about naming it `promote_version` so let's go for this one which is indeed better than the current name!

File: lib/opam_solve.ml (hunk 3 above)
Comment (Leonidas-from-XIV) [reply lib/opam_solve.ml:182]: It is more like `move_version_to_front`, because that describes what it is doing.

It is fine to dislike `function`, I can see the point because sometimes naming the arguments really helps, but in such case it would be good if the argument was called something else than `l`.

File: lib/opam_solve.ml (hunk 2 above)
Comment (Leonidas-from-XIV) [reply lib/opam_solve.ml:181]: Oh, yeah, I meand `preferred_version` as the name of the named argument, not the name of the function. I think "promote" captures what the function is attempting to do pretty well!

Comment (NathanReb) [reply lib/opam_solve.ml:181]: I feel like if we use `promote_version` than naming the argument `~preferred_version` is a bit redundant.

File: cli/lock.ml (hunk 8 above)
Comment (NathanReb) [reply cli/lock.ml:566]: Yeah that's shorter and clearer! Let's change it!

File: lib/opam_solve.ml (hunk 6 above)
Comment (NathanReb) [reply lib/opam_solve.ml:224]: Yeah, I went for an option when I started this as I thought it would allow more optimisation to avoid running useless operations over and over when there's no preferred_versions but it was indeed not very bright as looking up an empty map is just as efficient as checking whether an option is `None` either way.

**Review by Leonidas-from-XIV — APPROVED** [review]
Nice, let's merge it!
